pkg/sql/reassign_owned_by.go

```diff
@@ -13,15 +13,114 @@ package sql
+	// TODO(angelaw): Refactor the error checking out to authorization?
+	currentUser := p.SessionData().User
+
+	memberOf, err := p.MemberOfWithAdminOption(ctx, currentUser)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, oldRole := range n.OldRoles {
+		roleExists, err := p.RoleExists(ctx, oldRole)
+		if err != nil {
+			return nil, err
+		}
+
+		if !roleExists {
+			return nil, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", oldRole)
+		}
+
+		if _, ok := memberOf[oldRole]; !ok {
+			return nil, pgerror.Newf(
+				// TODO(angelaw): find error message from psql
+				pgcode.InsufficientPrivilege, "current user must be member of role %q", oldRole)
+		}
+	}
+
+	// Check new role exists
+	roleExists, err := p.RoleExists(ctx, n.NewRole)
+	if err != nil {
+		return nil, err
+	}
+
+	if !roleExists {
+		return nil, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", n.NewRole)
+	}
+
+	// And that current user has membership on new role (PostgresQL requirement)
+	if _, ok := memberOf[n.NewRole]; !ok {
+		return nil, pgerror.Newf(
+			// TODO(angelaw): find error message from psql
+			pgcode.InsufficientPrivilege, "current user must be member of role %q", n.NewRole)
+	}
+
+	// Find all objects in current database - use getAllDescriptors, newInternalLookupCtxFromDescriptors
+	allDescs, err := p.Descriptors().GetAllDescriptors(ctx, p.txn)
+	if err != nil {
+		return nil, err
+	}
+
+	currentDatabase := p.CurrentDatabase()
+	currentDbDesc, err := p.ResolveMutableDatabaseDescriptor(ctx, currentDatabase, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(angelaw): This is messy but we need an immutableDbDesc for description and
+	// we only have a ResolveMutableDatabaseDescriptor — any better ideas?
+	lCtx := newInternalLookupCtx(ctx, allDescs, currentDbDesc.ImmutableCopy().(*dbdesc.Immutable))
+
+	// For each role in old roles
+	for _, oldRole := range n.OldRoles {
+		// Iterate through each object in current DB
+		for _, dbDesc := range lCtx.dbDescs {
+			// We only expect 1 matching db (current database)
+			if p.IsOwner(dbDesc, oldRole) {
+				p.AlterDatabaseOwner(ctx, tree.AlterDatabaseOwner) // Does not include new owner string
+			}
+		}
+		for _, schemaDesc := range lCtx.schemaDescs {
+			if p.IsOwner(schemaDesc, oldRole) {
+				// What is job desc here?
+				p.alterSchemaOwner(ctx, dbDesc, schemaDesc, n.NewRole, jobDesc)
+			}
+		}
+
+		for _, tbDesc := range lCtx.tbDescs {
+			if p.IsOwner(tbDesc, oldRole) {
+				// Need to populate alterTableNode?
+				p.alterTableOwner(ctx, alterTableNode, n.NewRole)
+			}
+		}
+
+		for _, typDesc := range lCtx.typDescs {
+			if p.IsOwner(typDesc, oldRole) {
+				// Need to populate alterTypeNode?
+				p.alterTypeOwner(ctx, alterTypeNode, n.NewRole)
+			}
+		}
+	}
+
+	// Not sure we can directly call those methods as checkCanAlterToNewOwner is called and has different checks
+	// I think it actually works fine as long as we pass in hasOwnership = true
```

Comment:
Would appreciate a gut check here on whether it's okay that we call methods that call https://github.com/cockroachdb/cockroach/blob/0b105ec147985b65a2331784d8aadc0a1eec44cb/pkg/sql/authorization.go#L552 (`checkCanAlterToNewOwner`) with a `true` value for `hasOwnership` and not affect functionality